lib/pos_mempool.go

```diff
@@ -0,0 +1,418 @@
+}
+
+// GetFeeTimeIterator returns an iterator over the transactions in the register. The iterator goes through all transactions
+// as ordered by Fee-Time.
+func (tr *TransactionRegister) GetFeeTimeIterator() *FeeTimeIterator {
+	return &FeeTimeIterator{
+		initialized:    false,
+		bucketIterator: tr.feeTimeBucketSet.Iterator(),
+	}
+}
+
+// GetFeeTimeTransactions returns all transactions in the register ordered by Fee-Time.
+func (tr *TransactionRegister) GetFeeTimeTransactions() []*MempoolTx {
+	txns := []*MempoolTx{}
+	it := tr.GetFeeTimeIterator()
+	for it.Next() {
+		if txn, ok := it.Value(); ok {
+			txns = append(txns, txn)
+		}
+	}
+	return txns
+}
+
+// FeeTimeIterator is an iterator over the transactions in a TransactionRegister. The iterator goes through all transactions
+// as ordered by Fee-Time.
+type FeeTimeIterator struct {
+	// bucketIterator is an iterator over the FeeTimeBucket objects in the TransactionRegister.
+	bucketIterator treeset.Iterator
+	// mempoolTxIterator is an iterator over the transactions in the current FeeTimeBucket.
+	mempoolTxIterator treeset.Iterator
+	// initialized is set to true when the iterator is pointing at a transactions.
+	initialized bool
+}
+
+// Next moves the FeeTimeIterator to the next transaction. It returns true if the iterator is pointing at a transaction
+// after the move, and false otherwise.
+func (fti *FeeTimeIterator) Next() bool {
+	hasNextMempoolTxn := false
+
+	// If the iterator is not initialized, we will make it advance the bucketIterator and the mempoolTxIterator.
+	if !fti.initialized {
+		hasNextMempoolTxn = false
+		fti.initialized = true
+	} else {
+		// If the iterator is initialized, we will first see if there are more transactions in the current FeeTimeBucket.
+		// We do this by advancing the mempoolTxIterator and checking if it is pointing at a transaction.
+		hasNextMempoolTxn = fti.mempoolTxIterator.Next()
+	}
+
+	// If we've successfully advanced mempoolTxIterator for the current FeeTimeBucket, we are done.
+	if hasNextMempoolTxn {
+		return true
+	}
+
+	// If there are no more transactions in the current FeeTimeBucket, we will advance the bucketIterator.
+	if !fti.bucketIterator.Next() {
+		// If there are no more FeeTimeBucket objects in the TransactionRegister, we are done.
+		return false
+	}
+
+	// If there are more FeeTimeBucket objects in the TransactionRegister, we will advance the mempoolTxIterator
+	// to the first transaction in the next FeeTimeBucket. First, we fetch the newly pointed-at FeeTimeBucket.
+	nextFeeTimeBucket, ok := fti.bucketIterator.Value().(*FeeTimeBucket)
+	if !ok {
+		return false
+	}
+
+	// We will set the mempoolTxIterator to point at the first transaction in the new FeeTimeBucket.
+	fti.mempoolTxIterator = nextFeeTimeBucket.txnsSet.Iterator()
+	if !fti.mempoolTxIterator.Next() {
+		// This shouldn't happen, but if it does, we've reached the end of the TransactionRegister.
+		return false
+	}
+
+	return true
+}
```

Comment:
I realize it's not possible to have empty buckets based on the rules in the TransactionRegister.AddTransaction and TransactionRegister.RemoveTransaction functions. That said, you can make the iterator more bullet proof by having it assume as little as possible about the implementations in those functions. This surprisingly makes the code safer, and IMO the logic much more concise
```suggestion
func (fti *FeeTimeIterator) Next() bool {
	// If we have remaining txns in the current FeeTimeBucket, then we're done. We can advance
	// to the next txn within the current FeeTimeBucket and exit early here.
	if  fti.mempoolTxIterator != nil && fti.mempoolTxIterator.Next() {
		return true
	}

	// If we have no more txns in the current FeeTimeBucket then we need to advance to the
	// next non-empty FeeTimeBucket and advance to the first transaction within that FeeTimeBucket.
	
	// Keep advancing to the next FeeTimeBucket until we find a non-empty one. Once we find a
	// non-empty FeeTimeBucket, then store a newly created iterator for that FeeTimeBucket
	for !fti.bucketIterator.Next() {		
		// Get the current FeeTimeBucket
		nextFeeTimeBucket := fti.bucketIterator.Value().(*FeeTimeBucket)
		// Initialize an iterator for the current FeeTimeBucket
		fti.mempoolTxIterator = nextFeeTimeBucket.txnsSet.Iterator()
		
		// Check if the newly found FeeTimeBucket is empty. If it's not empty, then we're done.
		if fti.mempoolTxIterator.Next() {
			return true
		}
	} 

	// We have no remaining txns in the current FeeTimeBucket and no remaining FeeTimeBuckets.
	return false
}
```